googlebigtable2/src/main/java/site/ycsb/db/GoogleBigtable2Client.java

```diff
@@ -0,0 +1,420 @@
+  public Status scan(
+      String table,
+      String startkey,
+      int recordcount,
+      Set<String> fields,
+      Vector<HashMap<String, ByteIterator>> results) {
+    if (debug) {
+      System.out.printf("Doing scan for %s for %d records%n", startkey, recordcount);
+    }
+    Filters.Filter filter = buildFilter(fields);
+
+    Query query =
+        Query.create(table)
+            .filter(filter)
+            .range(Range.ByteStringRange.unbounded().startClosed(startkey))
+            .limit(recordcount);
+
+    final List<Row> rows;
+    try {
+      rows = client.readRowsCallable().all().call(query);
+    } catch (Exception e) {
+      if (debug) {
+        e.printStackTrace();
+      }
+      return Status.ERROR;
+    }
+
+    for (Row row : rows) {
+      HashMap<String, ByteIterator> rowResult = new HashMap<>();
+      rowToMap(row, rowResult);
+      results.add(rowResult);
+    }
+    return Status.OK;
+  }
+
+  private static void rowToMap(Row row, Map<String, ByteIterator> result) {
+    for (RowCell s : row.getCells()) {
+      result.put(s.getQualifier().toStringUtf8(), new ByteStringWrapper(s.getValue()));
+    }
+  }
+
+  @Override
+  public Status delete(String table, String key) {
+    if (debug) {
+      System.out.println("Doing delete for key: " + key);
+    }
+
+    if (clientSideBuffering) {
+      getBatcher(table).add(RowMutationEntry.create(key).deleteRow());
+      return Status.BATCHED_OK;
+    }
+
+    RowMutation m = RowMutation.create(TableId.of(table), key).deleteRow();
+    try {
+      client.mutateRow(m);
+    } catch (Exception e) {
+      if (debug) {
+        e.printStackTrace();
+      }
+      return Status.ERROR;
+    }
+
+    return Status.OK;
+  }
+
+  @Override
+  public Status insert(String table, String key, Map<String, ByteIterator> values) {
+    return update(table, key, values);
+  }
+
+  @Override
+  public Status update(String table, String key, Map<String, ByteIterator> values) {
+    if (debug) {
+      System.out.println("Setting up write for key: " + key);
+    }
+
+    try {
+      if (clientSideBuffering) {
+        return writeBatched(table, key, values);
+      } else {
+        return writeDirect(table, key, values);
+      }
+    } catch (Exception e) {
+      if (debug) {
+        e.printStackTrace();
+      }
+      return Status.ERROR;
+    }
+  }
+
+  private Status writeDirect(String table, String key, Map<String, ByteIterator> values) {
+    RowMutation m = RowMutation.create(TableId.of(table), key);
+    mapToMutation(values, m);
+
+    try {
+      client.mutateRow(m);
+    } catch (Exception e) {
+      if (debug) {
+        e.printStackTrace();
+      }
+      return Status.ERROR;
+    }
+    return Status.OK;
+  }
+
+  private Status writeBatched(String table, String key, Map<String, ByteIterator> values) {
+    RowMutationEntry m = RowMutationEntry.create(key);
+    mapToMutation(values, m);
+    getBatcher(table).add(m);
+
+    return Status.BATCHED_OK;
```

Comment:
If there are individual failures should we still return OK status?

Reply:
I dont think ycsb has any way to deal with batch entry failures :(